### ISSUE TITLE
Issue connecting to NDI Polaris on macOS

### ISSUE THREAD

**Issue by olevs** [PlusToolkit/PlusLib#263]
I just tried connecting to NDI Polaris Spectra using this file unmodified: PlusLibData/ConfigFiles/PlusDeviceSet_Server_NDIPolaris.xml

I haven't looked much into it yet, so it may be that I have to change some parameter in the xml file. I just thought I should let you know.
I'm using mscOS 10.12.6 (Sierra)
I verified that I get tracking with CustusX on the same machine, so the Polaris should be connected correctly.

./PlusServer --config-file=/Users/olevs/dev/plus_rel/PlusB-bin/PlusLibData/ConfigFiles/PlusDeviceSet_Server_NDIPolaris.xml
System start timestamp: 1.50972e+09
Software version: Plus-2.5.0.1e82ad62 - Mac
Logging at level 3 (INFO) to file: /Users/olevs/dev/plus_rel/PlusB-bin/bin/Output/110317_144533_PlusLog.txt
Server status: Reading configuration.
Server status: Connecting to devices.
|WARNING|023.459349| Unable to set requested baud rate. Reverting to auto-select.| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx(339)
|ERROR|023.517388| Incorrect number of command parameters| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx(357)
|ERROR|023.517948| Unable to connect to NDI device.| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx(302)
|ERROR|023.518020| TrackerDevice: Cannot connect to data source, ConnectInternal failed| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1204)
|ERROR|023.518058| Unable to connect device: TrackerDevice.| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(331)
|ERROR|023.520984| Datacollector failed to connect to devices| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusServer/Tools/PlusServer.cxx(129)
[14:45:57 ~/dev/plus_rel/PlusB-bin/bin]

**Comment by adamrankin** [comment]
I will hook up my spectra to our mac machine and diagnose it this weekend.

**Comment by adamrankin** [comment]
Hey, could you help me connect to my Spectra system on Mac? I've plugged it in, installed Toolbox, but I have no /dev/tty.XYZ entry. I am using the NDI Host Converter USB

**Comment by olevs** [comment]
Try installing the FTDI USB driver linked to from https://www.custusx.org/uploads/user_doc/nightly/install_ndi_driver.html

The NDI installer should also include this installer (a slightly older version, but it probably works fine). This should be located in the folder "/Applications/NDI ToolBox/USB Driver".

**Comment by olevs** [comment]
I get the exact same error when trying to connect PlusServer to NDI Polaris Spectra on Ubuntu 16.04.

**Comment by olevs** [comment]
I played a bit around with the code in vtkPlusNDITracker::InternalConnectSerial()

Looking at the IGSTK code,  used by CustusX I see that "RESET " and "INIT " is called before "COMM ...". Actually I think IGSTK may send a serial communication break signal instead of RESET, but calling "RESET " should also work, and it causes the Polaris to beep twice, but the COMM command still fails with error code 07 (= Incorrect number of parameters).

It looks like the COMM command in line 331
https://github.com/PlusToolkit/PlusLib/blob/88b862990b2a786d8b051ae45fbf1e2c87574ccd/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx#L331
and line 352:
https://github.com/PlusToolkit/PlusLib/blob/88b862990b2a786d8b051ae45fbf1e2c87574ccd/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx#L352
may be slightly wrong?

I tried replacing the %X with %d, but the command still fails.

**Comment by adamrankin** [comment]
I think that a serial break should be sent, I'll double check things when I get this *!%ing thing to show up under Mac.

**Comment by adamrankin** [comment]
> Note that serial break must only be sent if you cannot communicate with the device and you want to reset the baud rate. Polaris beeping means error, it should be avoided.

This is not the case, the NDI Track software sends a serial break by default. Beeping simply means the device is reset.

> In a shared environment, where different the same tracker is used by several projects, using different baud rates, the current approach of switching back to 9600 baud on disconnect is probably more robust.

I disagree, sending a serial break on connect guarantees that the device is at 9600 baud. If other software is used, or if Plus crashes, then the device is left in an unknown state.

The ndicapi project does a fair bit of error managing on Windows, but is not as thorough on other operating systems. This is perhaps why we haven't noticed any problems, as other OSes are not in as common of usage.

**Comment by lassoan** [comment]
My suggestion is to try to communicate with the device at the expected baud rate (in case of Plus, at 9600). If that fails, _then_ send the serial break and INIT.

It is true that beep does not strictly mean error, but it is for alerting the users for major events (power on, communication timeout, communication reset). Beep is loud and distracting if it happens too frequently, and may mask errors alerts (they sound the same).

INIT takes several seconds, so we should not force doing it unless it is really needed. If you only connect to a tracker once during a procedure then 5-second delay does not matter, but if reconnecting to devices is part of the workflow then 5 seconds is a long time.

> NDI Track software sends a serial break by default

I guess that's a diagnostic/test application. Unnecessary beeps and additional 5 second delay to connection setup may not matter that much when you troubleshoot your tracker. We can have an option to force send serial break/INIT, but in Plus I would not enable this by default.

**Comment by adamrankin** [comment]
> I just tried connecting PlusServer to NDI Polaris Spectra on Ubuntu 16.04, and got exactly same error.

The Mac and Linux code is essentially identical, when I fix one, it'll 99% fix the other.

**Comment by adamrankin** [comment]
Connected to tracker in Mac using NDI Track, working on Plus now.

**Comment by adamrankin** [comment]
Ok, fixes committed, would you be able to test? I was able to receive data from our Spectra connected to a Mac.

**Comment by olevs** [comment]
Nice. I will try to test this on monday.

**Comment by olevs** [comment]
I got the Ubuntu 16.04 setup working.

Did't have much time testing the Mac setup, but this still fails.

**Comment by adamrankin** [comment]
Ok, could you pass me an error log?

**Comment by olevs** [comment]
I wanted to do a bit more testing before sending an error log. Managed to connect to Polaris today. I used the same file as I used on Ubuntu, and just had to remove the SerialPort reference from my xml file to make it work.

**Comment by adamrankin** [comment]
Ok, it fails with SerialPort set?

**Comment by olevs** [comment]
Yes, the xml file I was using had SerialPort="3". The log said: "Failed to detect device on serial port 3...", so that should be an ok behavior. I don't think macOS use the same kind of numbering

**Comment by adamrankin** [comment]
Yes, but serial port 1 should work, if for example you only have one device plugged in, internally that should find /dev/cu.ttyXYZ

**Comment by olevs** [comment]
Setting SerialPort="1" also fail:

Server status: Connecting to devices.
|ERROR|007.692741| Failed to detect device on serial port 1. Error opening serial connection| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx(327)
|ERROR|007.692813| Unable to connect to NDI device.| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/NDICAPITracking/vtkPlusNDITracker.cxx(306)
|ERROR|007.692848| PolarisDevice: Cannot connect to data source, ConnectInternal failed| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1204)
|ERROR|007.692878| Unable to connect device: PolarisDevice.| in /Users/olevs/dev/plus_rel/PlusB-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(331)

On my Mac Polaris appears as: /dev/tty.usbserial-00005114

**Comment by lassoan** [comment]
In the meantime, we'll create the new stable release based on the current code. We can create patch release later if you come up with a fix.

**Comment by adamrankin** [comment]
I was able to connect using SerialPort=2, serial port 1 maps to /dev/cu.bluetoothXYZ, serial port 2 maps to /dev/cu.usbserial-XYZ

It basically iterates over all entries in /dev looking for cu.*, and returning index "SerialPort-1".

I don't love the serial port entry in the configuration, but there's no clean way to describe the different behavior for different OSs in a variable name.

**Comment by lassoan** [comment]
I would much prefer to set the actual device name (COM1, COM2, COM10, ... on Windows) rather than just some number (1, 2, 10, ...).
Number of course may be still needed for autodetect mode, so that the program can automatically iterate through the devices using a for loop (but it may not even be necessary if you can query the list of available device names).

**Comment by adamrankin** [comment]
The number is not needed for autodetect. Autodetect just scans all serial ports of as reported by the OS (well, ndicapi, which does OS specific serial port names).

I would much prefer the actual device name OS specific choice ("COM4", "/dev/cu.usbserial-XYZ", etc...)

**Comment by adamrankin** [comment]
However, the autodetect seems quite reliable, it's worked for me on both Windows and Mac, and I am testing Linux now.

When compiled with VS2012+, all serial ports are checked in parallel, reducing wait time.

**Comment by adamrankin** [comment]
Optionally, this code could be made cross-platform by changing to std::future and std::async

**Comment by lassoan** [comment]
If we all agree, then we probably should deprecate referring to ports as "1", "2", ... and use actual device names instead.

**Comment by adamrankin** [comment]
Confirmed working on Linux as well, both autodetect and specified.
